File: src/connect4/rules.py
"""
Game rules and mechanics for Connect 4 with Removals.

Includes:
- Legal move generation
- Move application (placement and removal with gravity)
- Win detection
- Terminal state detection
"""

from typing import List, Optional, Tuple
from .board import GameState
from .moves import PlacementMove, RemovalMove


def get_legal_moves(state: GameState) -> List:
    """
    Generate all legal moves for the current player.

    Returns a list of PlacementMove and/or RemovalMove objects.
    """
    moves = []

    # Placement moves: any non-full column
    for col in range(GameState.COLS):
        if not state.is_column_full(col):
            moves.append(PlacementMove(column=col))

    # Removal moves: any occupied cell (if player has removals left)
    if state.removals_remaining[state.current_player] > 0:
        for row in range(GameState.ROWS):
            for col in range(GameState.COLS):
                if state.is_cell_occupied(row, col):
                    moves.append(RemovalMove(row=row, col=col))

    return moves


def apply_move(state: GameState, move) -> GameState:
    """
    Apply a move to the state and return the updated state.

    Handles:
    - Placement: drop piece to lowest available row in column
    - Removal: remove piece and apply gravity
    - Win detection
    - Terminal state detection
    - Player switching
    """
    new_state = state.copy()
    new_state.move_history.append(move)

    if isinstance(move, PlacementMove):
        _apply_placement(new_state, move)
    elif isinstance(move, RemovalMove):
        _apply_removal(new_state, move)

    # Check for win
    if _check_win(new_state):
        new_state.winner = state.current_player
        new_state.is_terminal = True
    # Check for draw (board full)
    elif _is_board_full(new_state):
        new_state.is_terminal = True
        new_state.winner = None

    # Switch player if game is not terminal
    if not new_state.is_terminal:
        new_state.current_player = 1 - new_state.current_player

    return new_state


def _apply_placement(state: GameState, move: PlacementMove) -> None:
    """Apply a placement move: drop piece in column to lowest empty row."""
    col = move.column
    # Find the first empty row from the bottom
    for row in range(GameState.ROWS - 1, -1, -1):
        if state.get_cell(row, col) is None:
            state.set_cell(row, col, state.current_player)
            return
    # Should never reach here if move is legal
    raise ValueError(f"Column {col} is full but move was marked legal")


def _apply_removal(state: GameState, move: RemovalMove) -> None:
    """
    Apply a removal move: remove piece at (row, col) and apply gravity.

    Gravity: all pieces in the same column above the removed piece fall down by 1.
    """
    row, col = move.row, move.col

    # Remove the piece
    if state.get_cell(row, col) is None:
        raise ValueError(f"Cell ({row}, {col}) is empty but removal was marked legal")

    state.set_cell(row, col, None)

    # Apply gravity: shift pieces above the gap downward
    # Iterate from just above the gap (row-1) upward to the top (0)
    for source_row in range(row - 1, -1, -1):
        piece = state.get_cell(source_row, col)
        if piece is not None:
            state.set_cell(source_row + 1, col, piece)
            state.set_cell(source_row, col, None)

    # Decrement removal count for current player
    state.removals_remaining[state.current_player] -= 1


def _check_win(state: GameState) -> bool:
    """
    Check if the current player has won (4 in a row).

    Returns True if current player has created a 4-in-a-row.
    """
    player = state.current_player

    # Check horizontal
    for row in range(GameState.ROWS):
        for col in range(GameState.COLS - 3):
            if all(state.get_cell(row, col + i) == player for i in range(4)):
                return True

    # Check vertical
    for col in range(GameState.COLS):
        for row in range(GameState.ROWS - 3):
            if all(state.get_cell(row + i, col) == player for i in range(4)):
                return True

    # Check diagonal (top-left to bottom-right)
    for row in range(GameState.ROWS - 3):
        for col in range(GameState.COLS - 3):
            if all(state.get_cell(row + i, col + i) == player for i in range(4)):
                return True

    # Check diagonal (top-right to bottom-left)
    for row in range(GameState.ROWS - 3):
        for col in range(3, GameState.COLS):
            if all(state.get_cell(row + i, col - i) == player for i in range(4)):
                return True

    return False


def _is_board_full(state: GameState) -> bool:
    """Check if the board is completely full (draw condition)."""
    for col in range(GameState.COLS):
        if not state.is_column_full(col):
            return False
    return True


def has_legal_moves(state: GameState) -> bool:
    """Check if there are any legal moves remaining."""
    return len(get_legal_moves(state)) > 0

File: engine/rules.c
#include "rules.h"

int column_height(const GameState *state, int col) {
    int h = 0;
    for (int r = 0; r < ROWS; r++) {
        if (state->board[r][col] != EMPTY) h++;
    }
    return h;
}

int is_legal_move(const GameState *state, Move move) {
    if (state->is_terminal) return 0;

    if (move.type == MOVE_PLACEMENT) {
        if (move.column < 0 || move.column >= COLS) return 0;
        return state->board[ROWS - 1][move.column] == EMPTY;
    }

    if (move.type == MOVE_REMOVAL) {
        if (state->removals_remaining[state->current_player] <= 0) return 0;
        if (move.row < 0 || move.row >= ROWS) return 0;
        if (move.col < 0 || move.col >= COLS) return 0;
        return state->board[move.row][move.col] != EMPTY;
    }

    return 0;
}

int check_win(const GameState *state, int player) {
    for (int r = 0; r < ROWS; r++) {
        for (int c = 0; c <= COLS - 4; c++) {
            if (state->board[r][c] == player &&
                state->board[r][c+1] == player &&
                state->board[r][c+2] == player &&
                state->board[r][c+3] == player) return 1;
        }
    }

    for (int c = 0; c < COLS; c++) {
        for (int r = 0; r <= ROWS - 4; r++) {
            if (state->board[r][c] == player &&
                state->board[r+1][c] == player &&
                state->board[r+2][c] == player &&
                state->board[r+3][c] == player) return 1;
        }
    }

    for (int r = 0; r <= ROWS - 4; r++) {
        for (int c = 0; c <= COLS - 4; c++) {
            if (state->board[r][c] == player &&
                state->board[r+1][c+1] == player &&
                state->board[r+2][c+2] == player &&
                state->board[r+3][c+3] == player) return 1;
        }
    }

    for (int r = 3; r < ROWS; r++) {
        for (int c = 0; c <= COLS - 4; c++) {
            if (state->board[r][c] == player &&
                state->board[r-1][c+1] == player &&
                state->board[r-2][c+2] == player &&
                state->board[r-3][c+3] == player) return 1;
        }
    }

    return 0;
}

int is_board_full(const GameState *state) {
    for (int c = 0; c < COLS; c++) {
        if (state->board[ROWS - 1][c] == EMPTY) return 0;
    }
    return 1;
}

static void apply_gravity_column(GameState *state, int col) {
    int write = 0;
    for (int r = 0; r < ROWS; r++) {
        if (state->board[r][col] != EMPTY) {
            int piece = state->board[r][col];
            state->board[r][col] = EMPTY;
            state->board[write][col] = piece;
            write++;
        }
    }
}

void apply_move(GameState *state, Move move) {
    int mover = state->current_player;

    if (move.type == MOVE_PLACEMENT) {
        for (int r = 0; r < ROWS; r++) {
            if (state->board[r][move.column] == EMPTY) {
                state->board[r][move.column] = mover;
                break;
            }
        }
    } else {
        state->board[move.row][move.col] = EMPTY;
        apply_gravity_column(state, move.col);
        state->removals_remaining[mover]--;
    }

    int opponent = (mover == PLAYER_X) ? PLAYER_O : PLAYER_X;
    if (check_win(state, mover)) {
        state->is_terminal = 1;
        state->winner = mover;
    } else if (check_win(state, opponent)) {
        state->is_terminal = 1;
        state->winner = opponent;
    } else if (is_board_full(state)) {
        state->is_terminal = 1;
        state->winner = 0;
    }

    state->current_player = opponent;
}
